pkg/validation/validation_test.go

```diff
@@ -0,0 +1,480 @@
+			Spec: api.WorkflowSpec{
+				Steps: map[string]api.WorkflowStep{
+					"one": {
+						JobTemplate: &k8sBatch.JobTemplateSpec{},
+						ExternalRef: &k8sApi.ObjectReference{},
+					},
+				},
+				Selector: &k8sApiUnv.LabelSelector{
+					MatchLabels: map[string]string{"a": "b"},
+				},
+			},
+		},
+	}
+
+	for k, v := range errorCases {
+		errs := ValidateWorkflow(&v)
+		if len(errs) == 0 {
+			t.Errorf("expected failure for %s", k)
+		} else {
+			s := strings.Split(k, ":")
+			err := errs[0]
+			if err.Field != s[0] || !strings.Contains(err.Error(), s[1]) {
+				t.Errorf("unexpected error: %v, expected: %s", err, k)
+			}
+		}
+	}
+}
+
+func NewWorkflow() api.Workflow {
+	return api.Workflow{
+		ObjectMeta: k8sApi.ObjectMeta{
+			Name:            "mydag",
+			Namespace:       k8sApi.NamespaceDefault,
+			UID:             k8sTypes.UID("1uid1cafe"),
+			ResourceVersion: "42",
+		},
+		Spec: api.WorkflowSpec{
+			Steps: map[string]api.WorkflowStep{
+				"one": {
+					JobTemplate: &k8sBatch.JobTemplateSpec{},
+				},
+				"two": {
+					JobTemplate: &k8sBatch.JobTemplateSpec{},
+				},
+			},
+			Selector: &k8sApiUnv.LabelSelector{
+				MatchLabels: map[string]string{"a": "b"},
+			},
+		},
+		Status: api.WorkflowStatus{
+			StartTime: &k8sApiUnv.Time{Time: time.Date(2009, time.January, 1, 27, 6, 25, 0, time.UTC)},
+			Statuses: map[string]api.WorkflowStepStatus{
+				"one": {
+					Complete: false,
+				},
+				"two": {
+					Complete: false,
+				},
+			},
+		},
+	}
+}
+
+func TestValidateWorkflowUpdate(t *testing.T) {
+
+	type WorkflowPair struct {
+		current      api.Workflow
+		patchCurrent func(*api.Workflow)
+		update       api.Workflow
+		patchUpdate  func(*api.Workflow)
+	}
+	errorCases := map[string]WorkflowPair{
+		"metadata.resourceVersion: Invalid value: \"\": must be specified for an update": {
+			current: NewWorkflow(),
+			update:  NewWorkflow(),
+			patchUpdate: func(w *api.Workflow) {
+				w.ObjectMeta.ResourceVersion = ""
+			},
+		},
+		"workflow: Forbidden: cannot update completed workflow": {
+			current: NewWorkflow(),
+			patchCurrent: func(w *api.Workflow) {
+				s1 := w.Status.Statuses["one"]
+				s1.Complete = true // one is complete
+				w.Status.Statuses["one"] = s1
+				s2 := w.Status.Statuses["two"]
+				s2.Complete = true // two is complete
+				w.Status.Statuses["two"] = s2
+			},
+			update: NewWorkflow(),
+		},
+		"spec.steps: Forbidden: cannot delete running step \"one\"": {
+			current: NewWorkflow(),
+			patchCurrent: func(w *api.Workflow) {
+				delete(w.Status.Statuses, "two") // one is running
+			},
+			update: NewWorkflow(),
+			patchUpdate: func(w *api.Workflow) {
+				// we delete "one"
+				delete(w.Spec.Steps, "one") // trying to remove a running step
+				delete(w.Status.Statuses, "two")
+			},
+		},
+		"spec.steps: Forbidden: cannot modify running step \"one\"": {
+			current: NewWorkflow(),
+			patchCurrent: func(w *api.Workflow) {
+				delete(w.Status.Statuses, "two") // one is running
```

Comment:
Why does it say `one is running`? Shouldn't that be `two is running`? Or something like `two is now not running anymore`